### ISSUE TITLE
MSTP instance configuration cleanup on protocol version change

### ISSUE THREAD

**Issue by sonideepa** [mstpd/mstpd#49]
When protocol is changed using force protocol version command from mstp to stp or rstp, it is observed that all mstp instance configuration is not cleaned up. As  multiple instance configuration is not valid when protocol is STP or RSTP. Can you please let us know any reason to retain instance configuration.

Thanks!

**Comment by sonideepa** [comment]
Please find below sequence of mstpctl command, created instance not cleanup after changing protocol to STP
# ./mstpctl createtree br0 1
#./mstpctl showtree br0 1
virbr0 MSTI 1 info
  bridge id          8.001.52:54:00:1F:38:80
  regional root      8.001.52:54:00:1F:38:80
  root port          none
  internal path cost 0
  time since topology change 17
  topology change count      0
  topology change            no
  topology change port       None
  last topology change port  None
[root@localhost mstpd-master]# ./mstpctl setforcevers virbr0 stp
[root@localhost mstpd-master]# ./mstpctl showtree virbr0 1
virbr0 MSTI 1 info
  bridge id          8.001.52:54:00:1F:38:80
  regional root      8.001.52:54:00:1F:38:80
  root port          none
  internal path cost 0
  time since topology change 30
  topology change count      0
  topology change            no
  topology change port       None
  last topology change port  None

Thanks!!!

**Comment by commodo** [comment]
I'll take a look.
I'm a little swamped with other work.

Typically, `mstpctl setforcevers virbr0 stp` should be done before any configuration; at least that's how we use it.
Maybe we should document this somewhere.

But I agree it makes sense to reset data when switching between protocols.

**Comment by sonideepa** [comment]
Thanks for update!!!

**Comment by mg964** [comment]
Picking up on this thread... Having the same sort of issue, only I want to delete an MSTP instance. That's easy simply call deletetree:
```
vyatta@vm-mstp0-1:~$ /sbin/mstpctl showtree br0 1
br0 MSTI 1 info
  bridge id          7.001.42:C6:01:00:01:01
  regional root      7.001.42:C6:01:00:01:01
  root port          none
  internal path cost 0
  time since topology change 1183
  topology change count      4
  topology change            no
  topology change port       dp0p1s1
  last topology change port  dp0p1s2
vyatta@vm-mstp0-1:~$ 
vyatta@vm-mstp0-1:~$ /sbin/mstpctl deletetree br0 1
CTL_delete_msti: Got return code 0, -1
br0 Can't delete MSTID(1): there are FIDs allocated to it

vyatta@vm-mstp0-1:~$ 

```

**Comment by commodo** [comment]
hey ;

if the case is that you're also changing protocols when a bridge is created, i think that should not be a supported case ; and an error should be returned by the mstpctl tool ; 

so, i'll try to make some time to look into it and fix this ;

thanks
Alex

**Comment by mg964** [comment]
Hi Alex,

No, not trying to change the protocol as such - although that is one situation where you may want to remove all MSTIs. What I wanted to do was simply delete an MSTI; a change in configuration means that an MSTI is no longer needed.

Suppose I have this:

```
vyatta@vm-mstp-1:~$ sudo /sbin/mstpctl showvid2fid br0
br0 VID-to-FID allocation table:
  FID 0: 1-9,11-19,21-29,31-39,41-4094
  FID 1: 10,20
  FID 2: 30,40
vyatta@vm-mstp-1:~$ sudo /sbin/mstpctl showfid2mstid br0
br0 FID-to-MSTID allocation table:
  MSTID 0: 0,3-4095
  MSTID 1: 1
  MSTID 2: 2
vyatta@vm-mstp-1:~$ 
``` 

And then want to change the mapping so that only VLAN 99 is mapped to FID 1:

```
vyatta@vm-mstp-1:~$ sudo /sbin/mstpctl setvid2fid br0 1:99
vyatta@vm-mstp-1:~$ sudo /sbin/mstpctl showvid2fid br0
br0 VID-to-FID allocation table:
  FID 0: 1-9,11-19,21-29,31-39,41-98,100-4094
  FID 1: 10,20,99
  FID 2: 30,40
vyatta@vm-mstp-1:~$ 
```
So setvid2fid is really "addvid2fid", yet there is no corresponding "delvid2fid" command, i.e. there doesn't appear to be a way to change the mapping. OK so my solution is to delete the MSTI tree and re-create it with the new mapping. Except I can't delete an MSTI.

Would it make sense to include an optional parameter to deletetree that indicated a forced delete was required? Remove the tree together with any VLAN/FID mappings.

Cheers,
Mark

**Comment by daoyu09** [comment]
> Hi Alex,
> 
> No, not trying to change the protocol as such - although that is one situation where you may want to remove all MSTIs. What I wanted to do was simply delete an MSTI; a change in configuration means that an MSTI is no longer needed.
> 
> Further suppose I have this:
> 
> ```
> vyatta@vm-mstp-1:~$ sudo /sbin/mstpctl showvid2fid br0
> br0 VID-to-FID allocation table:
>   FID 0: 1-9,11-19,21-29,31-39,41-4094
>   FID 1: 10,20
>   FID 2: 30,40
> vyatta@vm-mstp-1:~$ sudo /sbin/mstpctl showfid2mstid br0
> br0 FID-to-MSTID allocation table:
>   MSTID 0: 0,3-4095
>   MSTID 1: 1
>   MSTID 2: 2
> vyatta@vm-mstp-1:~$ 
> ```
> 
> And then want to change the mapping so that only VLAN 99 is mapped to FID 1:
> 
> ```
> vyatta@vm-mstp-1:~$ sudo /sbin/mstpctl setvid2fid br0 1:99
> vyatta@vm-mstp-1:~$ sudo /sbin/mstpctl showvid2fid br0
> br0 VID-to-FID allocation table:
>   FID 0: 1-9,11-19,21-29,31-39,41-98,100-4094
>   FID 1: 10,20,99
>   FID 2: 30,40
> vyatta@vm-mstp-1:~$ 
> ```
> 
> So setvid2fid is really "addvid2fid", yet there is no corresponding "delvid2fid" command, i.e. there doesn't appear to be a way to change the mapping. OK so my solution is to delete the MSTI tree and re-create it with the new mapping. Except I can't delete an MSTI.
> 
> Would it make sense to include an optional parameter to deletetree that indicated a forced delete was required? Remove the tree together with any VLAN/FID mappings.
> 
> Cheers, Mark

I meet with this problem too. Have you resolved this problem? If so, could you please share the resolution with me? Thank you!

**Comment by Jolyk** [comment]
to daoyu09 and mg964

You can "delvid2fid" via assigning to fid 0 or mstid 0. And then deletetree.
```
mstpctl setfid2mstid br0 0:2
mstpctl setvid2fid br0 0:2-3

```
